core/src/main/scala/org/apache/spark/rpc/netty/Dispatcher.scala

```diff
@@ -0,0 +1,226 @@
+import org.apache.spark.network.client.RpcResponseCallback
+import org.apache.spark.rpc._
+import org.apache.spark.util.ThreadUtils
+
+private class RpcEndpointPair(val endpoint: RpcEndpoint, val endpointRef: NettyRpcEndpointRef)
+
+private[netty] class Dispatcher(nettyEnv: NettyRpcEnv) extends Logging {
+
+  private val endpointToInbox = new ConcurrentHashMap[RpcEndpoint, Inbox]()
+
+  // need a name to RpcEndpoint mapping so that we can delivery the messages
+  private val nameToEndpoint = new ConcurrentHashMap[String, RpcEndpointPair]()
+
+  private val endpointToEndpointRef = new ConcurrentHashMap[RpcEndpoint, RpcEndpointRef]()
+
+  // Track the receivers whose inboxes may contain messages.
+  private val receivers = new LinkedBlockingQueue[RpcEndpoint]()
+
+  @GuardedBy("this")
+  private var stopped = false
+
+  def registerRpcEndpoint(name: String, endpoint: RpcEndpoint): NettyRpcEndpointRef = {
+    val addr = new NettyRpcAddress(nettyEnv.address.host, nettyEnv.address.port, name)
+    val endpointRef = new NettyRpcEndpointRef(nettyEnv.conf, addr, nettyEnv)
+    synchronized {
+      if (stopped) {
+        throw new IllegalStateException("RpcEnv has been stopped")
+      }
+      if (nameToEndpoint.putIfAbsent(name, new RpcEndpointPair(endpoint, endpointRef)) != null) {
+        throw new IllegalArgumentException(s"There is already an RpcEndpoint called $name")
+      }
+      endpointToEndpointRef.put(endpoint, endpointRef)
+      val inbox = new Inbox(endpointRef, endpoint)
+      endpointToInbox.put(endpoint, inbox)
+      receivers.put(inbox.endpoint)
+    }
+    endpointRef
+  }
+
+  def getRpcEndpointRef(endpoint: RpcEndpoint): RpcEndpointRef = endpointToEndpointRef.get(endpoint)
+
+  def removeRpcEndpointRef(endpoint: RpcEndpoint): Unit = endpointToEndpointRef.remove(endpoint)
+
+  def getRpcEndpointRef(name: String): RpcEndpointRef = nameToEndpoint.get(name).endpointRef
+
+  // Should be idempotent
+  private def unregisterRpcEndpoint(name: String): Unit = {
+    val endpointPair = nameToEndpoint.remove(name)
+    if (endpointPair != null) {
+      val inbox = endpointToInbox.get(endpointPair.endpoint)
+      if (inbox != null) {
+        inbox.stop()
+        receivers.put(inbox.endpoint)
+      }
+    }
+  }
+
+  def stop(rpcEndpointRef: RpcEndpointRef): Unit = {
+    synchronized {
+      if (stopped) {
+        // This endpoint will be stopped by Distpatcher.stop() method.
+        return
+      }
+      unregisterRpcEndpoint(rpcEndpointRef.name)
+    }
+  }
+
+  /**
+   * Send a message to all registered [[RpcEndpoint]]s.
+   * @param message
+   */
+  def broadcastMessage(message: BroadcastMessage): Unit = {
+    val iter = endpointToInbox.values().iterator()
+    while (iter.hasNext) {
+      val inbox = iter.next()
+      postMessageToInbox(inbox, message)
+    }
+  }
+
+  def postMessage(message: RequestMessage, callback: RpcResponseCallback): Unit = {
+    val receiver = nameToEndpoint.get(message.receiver.name)
```

Comment:
if you have a method `getInbox(receiver)`, you can simplify this method and the next:

    val inbox = getInbox(message.receiver.name)
    if (inbox != null) {
      ...
    } else {
      ...
    }

And avoid the lone `return`, which I always find ugly.